### PR TITLE
Robot Diagnostics

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/extlibs/README.md [hunk 1]
```diff
@@ -1,2 +0,0 @@
-# Extlibs
-The extlibs folder contains code that was imported from other sources, so it may follow a different code convention. We aim to limit the integration with these extlibs to a small interface, so we may modify the imported code to make that happen. To make integration easier, we also may convert the code to use the Bazel build system.
```
Comment (itsarune) [inline src/extlibs/README.md:2]: why are we deleting this?

File: src/extlibs/hrvo/agent.cpp [hunk 2]
```diff
@@ -33,17 +33,17 @@ void Agent::update()
         new_velocity_ = new_velocity_.normalize(max_speed_);
     }
 
-    const Vector dv = new_velocity_ - velocity_;
-    if (dv.length() < max_accel_ * simulator_->getTimeStep() || dv.length() == 0.f)
+    const Vector dv            = new_velocity_ - velocity_;
+    double max_allowed_delta_v = 1000000;
```
Comment (itsarune) [inline src/extlibs/hrvo/agent.cpp]: fine, you can pull this into a constant if you want

File: src/extlibs/README.md (hunk 1 above)
Comment (itsarune) [reply src/extlibs/README.md:2]: nvm doesn't matter it's robocup

File: src/extlibs/hrvo/simulator.cpp [hunk 3]
```diff
@@ -312,6 +312,7 @@ Vector HRVOSimulator::getRobotVelocity(unsigned int robot_id) const
     auto hrvo_agent = getFriendlyAgentFromRobotId(robot_id);
     if (hrvo_agent.has_value())
     {
+        LOG(WARNING) << "VELOCITY: " << hrvo_agent.value()->getVelocity();
```
Comment (itsarune) [inline src/extlibs/hrvo/simulator.cpp]: delete if you don't need it

File: src/proto/message_translation/power_frame_msg.hpp [hunk 4]
```diff
@@ -116,21 +119,77 @@ TbotsProto_PowerStatus createNanoPbPowerStatus(float battery_voltage,
...
-        static_cast<uint8_t*>(serialized_proto.data()), serialized_proto.size());
-    if (!pb_decode(&pb_in_stream, TbotsProto_PowerControl_fields, &nanopb_control))
+    TbotsProto_PowerPulseControl nanopb_control =
+        TbotsProto_PowerPulseControl_init_default;
+    switch (google_control.chicker().chicker_command_case())
     {
-        throw std::runtime_error(
-            "Failed to decode serialized PowerControl msg to nanopb when converting google protobuf to nanopb");
+        case TbotsProto::PowerControl::ChickerControl::kKickSpeedMPerS:
+            // TODO: convert with redis
```
Comment (itsarune) [inline src/proto/message_translation/power_frame_msg.hpp:130]: add issue if you can

File: src/extlibs/hrvo/agent.cpp [hunk 5]
```diff
@@ -33,17 +33,17 @@ void Agent::update()
...
     {
         velocity_ = new_velocity_;
     }
     else
     {
         // Calculate the maximum velocity towards the preferred velocity, given the
         // acceleration constraint
-        velocity_ =
-            velocity_ + (max_accel_ * simulator_->getTimeStep()) * (dv / dv.length());
+        velocity_ = velocity_ + (max_allowed_delta_v) * (dv / dv.length());
```
Comment (nimazareian) [inline src/extlibs/hrvo/agent.cpp]: I think we should revert these changes

File: src/proto/world.proto [hunk 6]
```diff
@@ -29,6 +29,7 @@ message World
     required Team enemy_team      = 4;
     required Ball ball            = 5;
     required GameState game_state = 6;
+    // required uint64 sequence_number = 7;
```
Comment (LiCody) [inline src/proto/world.proto:32]: remove ?

Comment (itsarune) [inline src/proto/world.proto:32]: remove

File: src/software/backend/unix_simulator_backend.cpp [hunk 7]
```diff
@@ -68,6 +68,7 @@ void UnixSimulatorBackend::onValueReceived(TbotsProto::PrimitiveSet primitives)
 void UnixSimulatorBackend::onValueReceived(World world)
 {
     world_output->sendProto(*createWorld(world));
+    world_output->sendProto(*createWorld(world));
```
Comment (itsarune) [inline src/software/backend/unix_simulator_backend.cpp:71]: why are we sending this twice?

File: src/software/jetson_nano/ansible/playbooks/deploy_powerboard.yml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    - name: Flashing... (this will take a while on the first run)
+      shell: '/opt/tbotspython/bin/platformio run --disable-auto-clean -t nobuild -t upload -d ~/powerloop/power'
+      register: result
+
+    - name: Reset powerboard to finish flashing
+      ansible.builtin.pause:
+        prompt: "Press enter to continue"
+        echo: no
+
+    #- name: Reset powerboard to finish flashing
```
Comment (itsarune) [inline src/software/jetson_nano/ansible/playbooks/deploy_powerboard.yml:41]: delete uncommented things in this file if you're not using it

File: src/software/jetson_nano/ansible/playbooks/deploy_powerboard.yml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      register: result
+      args:
+        chdir: ~/
+
+    - name: Put the powerboard in bootloader mode
+      ansible.builtin.pause:
+        prompt: "Press enter to continue"
+        echo: no
+
+    #- name: Enter bootloader mode
```
Comment (LiCody) [inline src/software/jetson_nano/ansible/playbooks/deploy_powerboard.yml:27]: add TODO once ui board is fixed or whatever

File: src/software/jetson_nano/primitive_executor.h [hunk 10]
```diff
@@ -67,6 +72,11 @@ class PrimitiveExecutor
      */
     Vector getTargetLinearVelocity(const unsigned int robot_id,
                                    const Angle& curr_orientation);
+    Vector getTargetLinearVelocity(const TbotsProto::MovePrimitive& move_primitive,
+                                   const RobotState& robot_state);
+    double getTargetLinearSpeed(const TbotsProto::MovePrimitive& move_primitive,
+                                const RobotState& robot_state);
```
Comment (itsarune) [inline src/software/jetson_nano/primitive_executor.h:78]: docs

File: src/software/jetson_nano/services/network.cpp [hunk 11]
```diff
@@ -34,10 +36,24 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
 {
     std::scoped_lock<std::mutex> lock(primitive_set_mutex);
     primitive_set_msg = input;
+
+    // LOG(DEBUG) << "interpacket delay primitives: " <<
+    // input.time_sent().epoch_timestamp_seconds() - last_primitive_time;
```
Comment (LiCody) [inline src/software/jetson_nano/services/network.cpp:41]: remove debug logs

File: src/software/jetson_nano/services/network.cpp [hunk 12]
```diff
@@ -34,10 +36,24 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
...
+
+    last_primitive_time     = input.time_sent().epoch_timestamp_seconds();
+    static uint64_t counter = 0;
+    counter++;
+    // if (last_sequence_number != input.sequence_number() - 1)
```
Comment (LiCody) [inline src/software/jetson_nano/services/network.cpp:46]: remove debug logs

File: src/software/jetson_nano/services/network.cpp (hunk 11 above)
Comment (itsarune) [inline src/software/jetson_nano/services/network.cpp:41]: remove uncommented lines

File: src/software/jetson_nano/thunderloop.cpp [hunk 13]
```diff
@@ -143,7 +147,7 @@ void Thunderloop::runLoop()
                 }
             }
 
-            // If the world msg is new, update the internal buffer
+            //// If the world msg is new, update the internal buffer
```
Comment (itsarune) [inline src/software/jetson_nano/thunderloop.cpp:169]: two "//"

Comment (LiCody) [inline src/software/jetson_nano/thunderloop.cpp:169]: extra backslashes

File: src/extlibs/hrvo/simulator.cpp (hunk 3 above)
Comment (LiCody) [inline src/extlibs/hrvo/simulator.cpp]: remove debug

File: src/software/power/powerloop_main.cc [hunk 14]
```diff
@@ -33,31 +29,52 @@ uint8_t incomingByte;
...
 }
 
+
 void loop()
 {
+    if (Serial.available() > 0)
+    {
+        digitalWrite(23, HIGH);
+        timeout = 0;
+    }
```
Comment (LiCody) [inline src/software/power/powerloop_main.cc:65]: remove heartbeat

**Review by LiCody — COMMENTED** [review]
Really great changes to the power service and powerloop

File: src/software/jetson_nano/primitive_executor.cpp [hunk 15]
```diff
@@ -107,9 +193,20 @@ std::unique_ptr<TbotsProto::DirectControlPrimitive> PrimitiveExecutor::stepPrimi
         case TbotsProto::Primitive::kMove:
         {
             // Compute the target velocities
-            Vector target_velocity = getTargetLinearVelocity(robot_id, curr_orientation);
-            AngularVelocity target_angular_velocity =
-                getTargetAngularVelocity(current_primitive_.move(), curr_orientation);
+            // Vector target_velocity = getTargetLinearVelocity(current_primitive_.move(),
+            // robot_state);
```
Comment (itsarune) [inline src/software/jetson_nano/primitive_executor.cpp:195]: delete if you're not using

**Review by itsarune — DISMISSED** [review]
mostly minor issues, I don't particularly care if you fix them

they're more like notes for when it gets ported over to master
